lib/events.js

```diff
@@ -64,62 +64,72 @@ EventEmitter.prototype.getMaxListeners = function getMaxListeners() {
 // inside them can execute faster.
 function emitNone(handler, isFn, self) {
   if (isFn)
-    handler.call(self);
-  else {
-    var len = handler.length;
-    var listeners = arrayClone(handler, len);
-    for (var i = 0; i < len; ++i)
-      listeners[i].call(self);
+    return handler.call(self);
+  var len = handler.length, listeners = arrayClone(handler, len), me, listener;
+  for (var i = 0; i < len; ++i) {
+    listener = listeners[i]; me = self;
```

Comment:
there's a lot of duplicated logic in here... perhaps it's time to refactor this a bit?